qualtran/bloqs/arithmetic/permutation.py

```diff
@@ -0,0 +1,248 @@
+#  Copyright 2024 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+from functools import cached_property
+from typing import cast, Sequence, Set, TYPE_CHECKING, TypeAlias, Union
+
+from attrs import field, frozen
+
+from qualtran import (
+    Bloq,
+    bloq_example,
+    BoundedQUInt,
+    DecomposeTypeError,
+    QBit,
+    QUInt,
+    Signature,
+    Soquet,
+)
+from qualtran.bloqs.arithmetic.addition import XorK
+from qualtran.bloqs.arithmetic.comparison import EqualsAConstant
+from qualtran.linalg.permutation import (
+    CycleT,
+    decompose_permutation_into_cycles,
+    decompose_sparse_prefix_permutation_into_cycles,
+)
+from qualtran.symbolics import bit_length, is_symbolic, Shaped, slen, SymbolicInt
+
+if TYPE_CHECKING:
+    import sympy
+
+    from qualtran import BloqBuilder, SoquetT
+    from qualtran.resource_counting import BloqCountT, SympySymbolAllocator
+
+
+SymbolicCycleT: TypeAlias = Union[CycleT, Shaped]
+
+
+def _convert_cycle(cycle) -> Union[tuple[int, ...], Shaped]:
+    if isinstance(cycle, Shaped):
+        return cycle
+    return tuple(cycle)
+
+
+@frozen
+class PermutationCycle(Bloq):
+    r"""Apply a single permutation cycle on the basis states.
+
+    Args:
+        N: the total size the permutation acts on.
+        cycle: the permutation cycle to apply.
+
+    Registers:
+        x: integer register storing a value in [0, ..., N - 1]
+
+    References:
+        [A simple quantum algorithm to efficiently prepare sparse states](https://arxiv.org/abs/2310.19309v1)
+        Appendix B, Algorithm 7.
+    """
+
+    N: SymbolicInt
+    cycle: Union[tuple[int, ...], Shaped] = field(converter=_convert_cycle)
+
+    @cached_property
+    def signature(self) -> Signature:
+        return Signature.build_from_dtypes(x=BoundedQUInt(self.bitsize, self.N))
+
+    @cached_property
+    def bitsize(self):
+        return bit_length(self.N - 1)
```

Comment:
Will there ever be a case when the user wants to specify the `bitsize` instead of the default `log2(N)` ? We can make it an attribute and use this as a default value if the user doesn't specify anything; if you think that's relevant.